File: ProductOfArrayExcept.java
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
// Doing running product technique from initial index to last and the other way from the calculated array would give us the end result

class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] answer = new int[nums.length];
        int runningProduct = 1;
        // running Proudct is kept to trace the product of values till before curr index
        for(int i=1;i<nums.length;i++) {
            answer[i-1] = runningProduct;
            runningProduct *= nums[i-1];
        }
        // product for last pos is already here
        answer[nums.length -1] = runningProduct;
        runningProduct = 1;
        // doing running product from right to left in the answer array
        for(int i=nums.length - 1;i>=0;i--) {
            answer[i] *= runningProduct;
            runningProduct *= nums[i];
        }
        return answer;
    }
}

File: productofarrayexcept.py
#Time Complexity : O(n) where n -> lenegth of nums
#Space Complexity : O(1) as we are using rp and result is the output so not counted in space complexity
#did your code run on leetcode : yes
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        if nums == [] or len(nums) == 0: return []
        result = [0]*len(nums)
        
        rp = 1
        result[0] = 1
        
        #left pass
        for i in range(1,len(nums)):
            rp = rp * nums[i - 1]
            result[i] = rp

        rp = 1
        #right pass
        for i in reversed(range(0,len(nums)-1)):
            rp = rp * nums[i + 1]
            result[i] = result[i] * rp
        
        return result
            
